### PR TITLE
feat(rum-core): capture history.pushState calls as transactions

### PR DESCRIPTION
**Pull request by jahtalab**
This is part of the effort to [capture route change transactions](https://github.com/elastic/apm-agent-rum-js/issues/90). 

- patch history API
- refactoring of `TransactionService.startTransaction`
- introduce transaction reusability

### CODE REVIEW

File: packages/rum-core/src/common/patching/history-patch.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+export function patchHistory(callback) {
+  if (alreadyPatched || !window.history) {
+    return
+  }
+  alreadyPatched = true
+
+  const nativePushState = history.pushState
+  if (typeof nativePushState === 'function') {
+    history.pushState = function(state, title, url) {
```
Comment (vigneshshanmugam) [inline packages/rum-core/src/common/patching/history-patch.js]: I believe it's also called for redirects which should be ignored and not captured as part of transaction.

File: packages/rum-core/test/common/history-patch.spec.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    cancelFn()
+  })
+
+  beforeEach(function() {
+    events = []
+  })
+
+  it('should patch history.pushState', function() {
+    history.pushState(undefined, 'test', 'test')
+
```
Comment (vigneshshanmugam) [inline packages/rum-core/test/common/history-patch.spec.js]: Please add a e2e test as well with history.go and history.back methods and verifying the transactions with those names

File: packages/rum-core/src/performance-monitoring/performance-monitoring.js [hunk 3]
```diff
@@ -122,6 +123,12 @@ class PerformanceMonitoring {
           transactionService.removeTask(task.id)
         }
       }
+
+      if (event === INVOKE && task.source === HISTORY_PUSHSTATE) {
+        transactionService.startTransaction(task.data.title, 'route-change', {
```
Comment (vigneshshanmugam) [inline packages/rum-core/src/performance-monitoring/performance-monitoring.js]: The title can be empty and not sure if it's set all the time. we have to handle this case and I am not sure if it's set all the time during popstate events.

File: packages/rum-core/src/common/patching/history-patch.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import { INVOKE, HISTORY_PUSHSTATE } from '../constants'
+var alreadyPatched = false
+
+export function patchHistory(callback) {
```
Comment (vigneshshanmugam) [inline packages/rum-core/src/common/patching/history-patch.js]: It's safer to do this patch after we have captured the page load event transaction, if the user registers a state in history using 'pushState', we will accidently capture all our requests as part of route change transaction instead of page-load which will be a broken behaviour..

Would like to have rigorous E2E tests for this change.

File: packages/rum-core/src/performance-monitoring/performance-monitoring.js (hunk 3 above)
Comment (jahtalab) [reply packages/rum-core/src/performance-monitoring/performance-monitoring.js]: The title is not used in any of the frameworks I've seen. But we don't have to handle it here since the transaction title can be undefined and if it is then it would have 'unknown'.

File: packages/rum-core/src/common/patching/history-patch.js (hunk 4 above)
Comment (jahtalab) [reply packages/rum-core/src/common/patching/history-patch.js]: I considered this to be a feature, that we will capture the first route change as a part of the page-load transaction. This will give the user a complete perspective to the initial page load.

File: packages/rum-core/src/performance-monitoring/performance-monitoring.js (hunk 3 above)
Comment (vigneshshanmugam) [reply packages/rum-core/src/performance-monitoring/performance-monitoring.js]: Wont it be confusing then? If we are relying on title for the transaction name and multiple paths might lead to `Unknown` because of title being empty on different routes?

File: packages/rum-core/src/common/patching/history-patch.js (hunk 4 above)
Comment (vigneshshanmugam) [reply packages/rum-core/src/common/patching/history-patch.js]: I am really not sure if this a good feature to be honest, It would simply increase the number of transactions count on the details page plus instead of focusing on the transactions that matter the most, we are simply polluting the UI more less useful `redirected transactions`. 

For the first phase at-least, we can simply drop them and add it if there is a huge interest from users.

File: packages/rum-core/src/common/constants.js [hunk 5]
```diff
@@ -56,6 +61,11 @@ const RESOURCE_INITIATOR_TYPES = [
   'iframe'
 ]
 
+/**
+ * The amount of time it is allowed for a transaction to be reused in another startTransaction
+ */
+const REUSABILITY_THRESHOLD = 10000
```
Comment (vigneshshanmugam) [inline packages/rum-core/src/common/constants.js]: Any reasoning behind this particular value?

Comment (jahtalab) [reply packages/rum-core/src/common/constants.js]: I'll make the value into ms.

File: packages/rum-core/src/common/patching/history-patch.js (hunk 4 above)
Comment (jahtalab) [reply packages/rum-core/src/common/patching/history-patch.js]: I'll investigate this scenario more

Comment (jahtalab) [reply packages/rum-core/src/common/patching/history-patch.js]: @vigneshshanmugam , I will add e2e tests to #265 for this scenario

File: packages/rum-core/src/common/patching/history-patch.js (hunk 1 above)
Comment (jahtalab) [reply packages/rum-core/src/common/patching/history-patch.js]: I won't start a new transaction only redefines the old one which I think should be ok!

File: packages/rum-core/test/common/history-patch.spec.js (hunk 2 above)
Comment (jahtalab) [reply packages/rum-core/test/common/history-patch.spec.js]: Since the rum-core doesn't have e2e tests, will also add e2e tests to #265

File: packages/rum-core/src/common/patching/history-patch.js (hunk 4 above)
Comment (vigneshshanmugam) [reply packages/rum-core/src/common/patching/history-patch.js]: Sure, sounds good 👍

File: packages/rum-core/src/common/patching/history-patch.js (hunk 1 above)
Comment (vigneshshanmugam) [reply packages/rum-core/src/common/patching/history-patch.js]: Have to be careful here since if the redirect is too slow, we will discard it and create a new one. Its an edge case, but might want to think about reusability threshold.

**Comment by vigneshshanmugam** [comment]
Let's add todos for the other PR for the E2E Tests(edge cases) and also docs on supported vs unsupported libs.

**Comment by jahtalab** [comment]
I've added the todos to the other PR!
